### PR TITLE
Export Mininum Thin Pool Free Space through docker info

### PR DESCRIPTION
**Pull request by rhvgoyal**
Right now there is no way to know what's the minimum free space threshold
daemon is applying. It would be good to export it through docker info and
then user knows what's the current value. Also this could be useful to
higher level management tools which can look at this value and setup their
own internal thresholds for image garbage collection etc.

Signed-off-by: Vivek Goyal vgoyal@redhat.com

### CODE REVIEW

**Comment by rhvgoyal** [comment]
cc @thaJeztah

File: daemon/graphdriver/devmapper/driver.go [hunk 1]
```diff
@@ -74,6 +74,7 @@ func (d *Driver) Status() [][2]string {
 		{"Metadata Space Used", fmt.Sprintf("%s", units.HumanSize(float64(s.Metadata.Used)))},
 		{"Metadata Space Total", fmt.Sprintf("%s", units.HumanSize(float64(s.Metadata.Total)))},
 		{"Metadata Space Available", fmt.Sprintf("%s", units.HumanSize(float64(s.Metadata.Available)))},
+		{"Thin Pool Minimum Free Space", fmt.Sprintf("%v", s.MinFreeSpace)},
```
Comment (thaJeztah) [inline daemon/graphdriver/devmapper/driver.go]: is this only relevant if loopback devices are used? If so, perhaps it should only be shown in that case, like line 86 below

File: daemon/graphdriver/devmapper/deviceset.go [hunk 2]
```diff
@@ -2471,6 +2472,7 @@ func (devices *DeviceSet) Status() *Status {
 		}
 	}
 
+	status.MinFreeSpace = fmt.Sprintf("%v%%", devices.minFreeSpacePercent)
```
Comment (thaJeztah) [inline daemon/graphdriver/devmapper/deviceset.go]: Just a thought; Perhaps we should show both the absolute size (xx Gb) and percentage, something like `123 Gb (12%)`

Comment (thaJeztah) [inline daemon/graphdriver/devmapper/deviceset.go]: Actually, in the API, we shouldn't do any formatting, so either raw percentage, or calculated size

File: daemon/graphdriver/devmapper/driver.go (hunk 1 above)
Comment (rhvgoyal) [inline daemon/graphdriver/devmapper/driver.go]: No, this is relevant for all thin pool. It does not matter if thin pool gets created out of loop devices or regular block devices or if it is an lvm thin pool.

File: daemon/graphdriver/devmapper/deviceset.go (hunk 2 above)
Comment (rhvgoyal) [inline daemon/graphdriver/devmapper/deviceset.go]: @thaJeztah So existing format is fine? I am not sure what's the suggestion.

Or you are suggesting that we create two entries. One for percentage and one for absolute value.

Thin Pool Minimum Free Space Percent: X
Thin Pool Minimum Free Space: Y MB

Comment (thaJeztah) [inline daemon/graphdriver/devmapper/deviceset.go]: Open to suggestions; my train of thought was that "percentage" is a convenient way to define the amount of free space, but since all other values shown in `docker info` are absolute values, possibly the calculated value (`... MB`) is more easy to use in that context.

So, a bit on the fence on that one; perhaps someone else has suggestions

Comment (rhvgoyal) [inline daemon/graphdriver/devmapper/deviceset.go]: I am no the fence too. Displaying both percentage and absolute values are useful for human beings but if software is using this output, they can easily calculate one way or other.

At the same time, displaying both seems little redundant so I would rather avoid that and display either percentage or absolute value and let user and software calculate other if need be.

Given other values are absolute in docker info output, I think displaying absolute values for minimum free space is reasonable too. This will be fine if and when we support absolute values for dm.min_free_space parameter.

In summary, I think displaying either percentage or absolute both should be fine. But I am willing to implement absolute values for the sake of uniformity with other values.

Comment (thaJeztah) [inline daemon/graphdriver/devmapper/deviceset.go]: For consistency, I think we should go for absolute size then. Thanks for thinking along!

Comment (rhvgoyal) [inline daemon/graphdriver/devmapper/deviceset.go]: Ok, will change and push patch again.

**Comment by rhvgoyal** [comment]
@thaJeztah Updated PR. Now output is absolute value. Following is the sample.

# docker info
Containers: 5
Images: 2
Server Version: 1.11.0-dev
Storage Driver: devicemapper
 Pool Name: docker--vg-docker--pool
 Pool Blocksize: 524.3 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: 
 Metadata file: 
 Data Space Used: 370.1 MB
 Data Space Total: 2.999 GB
 Data Space Available: 2.629 GB
 Metadata Space Used: 180.2 kB
 Metadata Space Total: 8.389 MB
 Metadata Space Available: 8.208 MB
 Thin Pool Minimum Free Space: 599.8 MB
..
..
..

**Comment by thaJeztah** [comment]
Thanks! Looking good, I'll move this to code review

**Comment by calavera** [comment]
LGTM

**Comment by MHBauer** [comment]
Update to the example in the docs, at `docs/userguide/storagedriver/selectadriver.md` ?
[webpage](https://docs.docker.com/engine/userguide/storagedriver/selectadriver/)

**Comment by rhvgoyal** [comment]
@MHBauer Will do.

**Comment by rhvgoyal** [comment]
@MHBauer Updated the patch with update to the example in doc. PTAL.

**Comment by MHBauer** [comment]
LGTM

**Comment by cpuguy83** [comment]
LGTM
